_posts/2019-08-08-kubernetes-gated-deployments.md

```diff
@@ -0,0 +1,245 @@
+prolonged customer impact. Common techniques for assessing the impact of changes
+on key metrics of an application's functionality and performance include
+pre-post analysis, load testing and benchmarking, and canary deploys with manual
+inspection of metrics. There are several drawbacks with these approaches:
+
+* They are not scalable, and do not isolate and measure the impact of the change
+  itself: Manual assessments are time consuming and not scalable when teams are
+  making a large number of deployments throughout the day, which also makes it
+  challenging to determine the specific change that resulted in the impact on
+  key metrics.
+* They are prone to human error: Human analysis inherently leads to a subjective
+  decision that can result from error or bias, and it is difficult to
+  consistently determine if the metric deviations are due to random chance or
+  from a deployed change.
+* They are not representative of the system under realistic circumstances: For
+  example, engineers might choose to load test using synthetic models for user
+  requests, but it is difficult to ensure that those synthetic models produce a
+  traffic pattern similar enough to real production traffic to guarantee the
+  load test captures what engineers intend to measure.
+* It is difficult to monitor and respond to regressions in key metrics in a
+  timely manner: Teams may rely on monitors to show increased response times or
+  complaints from users, but at this point the change has already adversely
+  affected the customer experience, and engineers must roll back the change.
+
+## Kubernetes Gated Deployments
+
+To solve the problems described above, we’ve created Kubernetes Gated
+Deployments, which we use in place of manual analysis of metrics and
+intervention, enabling our engineering teams to run each deployment as a
+controlled experiment. Just as teams run experiments on product metrics to test
+client-side changes, they can use this Kubernetes extension to run experiments
+on infrastructure and code changes to assess the impact of those changes and
+detect issues by analyzing key metrics associated with the application's
+functionality and performance. This system automatically analyzes the metrics
+specified by engineers for the current and new versions of the application, and
+either rolls back the change or promotes it to take all production traffic.
+Kubernetes Gated Deployments does not require any application code changes, and
+relies on native Kubernetes resources to route traffic and manage the deployed
+application. It adds a new object type, called a `GatedDeployment`, that allows
+engineers to specify metric sources and analysis techniques, which the
+controller will use to compare versions. We designed this system to augment
+existing deployment processes by comparing the deployed versions to validate
+that the new one does no harm to the metrics specified in that object.
+
+## How it works
```

Comment:
I think there's an opportunity here to dig into the details of the developer experience here - what happens when there are multiple changes, or changes that depend on each other? What happens if I don't have a lot of traffic?

eg: 20 commits happen at about the same time; how long does it take for those changes to reliably get to prod through GatedDeployments, what happens if something goes wrong, and what's the likelihood that it identifies the commit with the defect?